src/docfx/lib/PathUtil.cs

```diff
@@ -0,0 +1,133 @@
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+
+namespace Microsoft.Docs
+{
+    /// <summary>
+    /// Provide utils of path
+    /// </summary>
+    public static class PathUtil
+    {
+        /// <summary>
+        /// Determines if a file or path is inside a folder. Input should both be normalized.
+        /// </summary>
+        public static bool Inside(string folder, string path)
+        {
+            Debug.Assert(folder.Length > 0 && folder[folder.Length - 1] == '/');
+
+            return folder == "./" || path.StartsWith(folder, StringComparison.OrdinalIgnoreCase);
+        }
+
+        /// <summary>
+        /// Create a relative path from one path to another file.
+        /// Use this over <see cref="Path.GetRelativePath(string, string)"/> when
+        /// <paramref name="fileRelativeTo"/> is a file.
+        /// </summary>
+        public static string GetRelativePathToFile(string fileRelativeTo, string path)
+        {
+            var directory = Path.GetDirectoryName(fileRelativeTo);
+            if (string.IsNullOrEmpty(directory))
+            {
+                return path;
+            }
+
+            return Path.GetRelativePath(directory, path);
+        }
+
+        /// <summary>
+        /// A normalized folder always ends with `/`, does not contain `\` and does not have consegative `.` or `/`.
+        /// </summary>
+        public static string NormalizeFolder(string path)
+        {
+            Debug.Assert(path.IndexOfAny(Path.GetInvalidPathChars()) < 0);
+
+            var str = Normalize(path);
+            if (str.Length == 0 || str == "/")
+            {
+                return "./";
+            }
+
+            if (!str.EndsWith('/'))
+            {
+                str += '/';
+            }
+            return str;
+        }
+
+        /// <summary>
+        /// A normalized file cannot end with `/`, does not contain `\` and does not have consegative `.` or `/`.
+        /// </summary>
+        public static string NormalizeFile(string path)
+        {
+            Debug.Assert(Path.GetDirectoryName(path).IndexOfAny(Path.GetInvalidPathChars()) < 0);
+            Debug.Assert(Path.GetFileName(path).IndexOfAny(Path.GetInvalidFileNameChars()) < 0);
+
+            return Normalize(path);
+        }
+
+        private static string Normalize(string path)
+        {
+            path = path.Replace('\\', '/');
+
+            var needReorder = false;
+            foreach (var c in path)
+            {
+                if (c == '.')
+                {
+                    needReorder = true;
+                    break;
+                }
+            }
+            if (!needReorder)
+            {
+                return path;
+            }
+
+            var parentCount = 0;
+            var rooted = path[0] == '/';
+            var stack = new List<string>();
+            foreach (var segment in path.Split('/', '\\'))
```

Comment:
>'\\' [](start = 52, length = 4)

remove?

Reply:
why?

---
In reply to: [179994083](https://github.com/dotnet/docfx/pull/2601#discussion_r179994083) [](ancestors = 179994083)